### PR TITLE
perf: Replace symmetric image stencil with anchored per-pair window for periodic image replication

### PR DESCRIPTION
**Pull request by n-gao**
This PR replaces the symmetric centered-stencil image replication with a pair-anchored window approach, fixing correctness on skewed triclinic cells and reducing over-replication. The benefits of this are purely performance and memory improvements.

## Image replication: anchored windows instead of symmetric stencils

Previously, `candidate_image_counts` computed a symmetric stencil of width `2 * ceil(ratio) + 1` centered at the origin, and `_generate_image_offsets` returned origin-centered coordinates. This over-counted images (e.g., ratio 0.8 produced 3 images/axis = 27 copies instead of the tight 2/axis = 8) and could miss the true closest image on highly skewed cells where round-based minimum-image shifts pick the wrong copy.

The new approach:

- `candidate_image_counts` returns `floor(2 * ratio) + 1` — the exact count of lattice planes a sphere of radius `cutoff` can reach along each axis.
- `_generate_image_offsets` now emits 0-based window coordinates (no origin centering), with the window center emitted first as a stable tie-break.
- `_get_candidate_images` anchors each candidate's window at `ceil(separation - ratio)`, which provably brackets every in-range periodic image of that pair. The `queries` table is now a required argument to compute per-pair separations.
- The `has_been_replicated` flag is removed. Instead, `_minimum_image_mask` uses a segment argmin over real distances to identify the closest copy per original candidate — correct even on acute/skewed cells where rounding fails.
- `replicate_for_images` no longer blends MIC shifts with image shifts; it passes the anchored offsets directly and delegates minimum-image detection entirely to `_minimum_image_mask`.

## Single-cell fast path in `_cell_list_subselect`

When `max_num_cells.size == 1`, every particle maps to the same cell and the stencil expansion is unnecessary. A new `single_cell` branch skips the vmap stencil tiling and the `jnp.unique` deduplication step, and passes `is_sorted=False` to `subselect` since the hashes are no longer pre-sorted by `unique`.

## Tests

- Updated `TestCandidateImageCounts`, `TestReplicateForImages`, `TestGetCandidateImagesIsFinite`, and `TestGenerateImageOffsets` to reflect the new window widths and 0-based coordinates.
- Added `test_single_cell_path_matches_multi_cell_path` verifying the single-cell fast path emits the same candidate set as the general path.
- Added `test_triclinic_images.py` with brute-force completeness and minimum-image flag tests across orthogonal, monoclinic, acute (35°), and anisotropic short-axis cells, exercising the full pipeline via `DenseNearestNeighborList` and `replicate_for_images` directly.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#213** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/213?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#215](https://github.com/cusp-ai-oss/kUPS/pull/215) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/215?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#216](https://github.com/cusp-ai-oss/kUPS/pull/216) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/216?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#212** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/212?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/212?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#211** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/211?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#210** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/210?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#204** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/204?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#209** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/209?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#203** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#202** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/202?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#201** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/201?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#186** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#185** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#183** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/183?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#182** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/182?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#181** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/181?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#180** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/180?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#178** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#206** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/206?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#205** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/kups/core/neighborlist/cell_list.py [hunk 1]
```diff
@@ -108,14 +108,23 @@ def _cell_list_subselect(
...
-    raw_shifted = jax.vmap(
-        lambda s: query_positions + s[None] / bins[queries.data.system]
-    )(stencil).reshape(-1, dim)
-    query_original = Index(
-        queries.keys, jnp.tile(jnp.arange(len(queries)), len(stencil))
-    )
+    # With a single cell per system the only reachable cell is "self", so a unit
+    # stencil suffices and every query maps to one cell -- the neighborhood dedup
+    # is a no-op and is skipped (subselect re-sorts via is_sorted=False).
+    single_cell = max_num_cells.size == 1
```
Comment (jonkhler) [inline src/kups/core/neighborlist/cell_list.py]: `is_single_cell`?

**Review by jonkhler — APPROVED** [review]

**Review by jonkhler — DISMISSED** [review]
var naming for readability

**Comment by n-gao** [comment]
> var naming for readability

​done :)

**Review by felixmusil — APPROVED** [review]
might need to increase the number of tests to make sure this does not have some corner case bugs

**Comment by n-gao** [comment]
### Merge activity

* **Jul 1, 8:54 PM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/212).
